File: src/prod/src/ServiceModel/DeployedNetworkQueryDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ServiceModel;
using namespace std;
using namespace Common;
using namespace Query;

StringLiteral const TraceComponent("DeployedNetworkQueryDescription");

DeployedNetworkQueryDescription::DeployedNetworkQueryDescription()
    : nodeName_()
    , queryPagingDescription_()
{
}

ErrorCode DeployedNetworkQueryDescription::FromPublicApi(FABRIC_DEPLOYED_NETWORK_QUERY_DESCRIPTION const &deployedNetworkQueryDescription)
{
    TRY_PARSE_PUBLIC_STRING(deployedNetworkQueryDescription.NodeName, nodeName_);    
    
    if (deployedNetworkQueryDescription.PagingDescription != NULL)
    {
        QueryPagingDescription pagingDescription;
        
        auto error = pagingDescription.FromPublicApi(*(deployedNetworkQueryDescription.PagingDescription));
        if (!error.IsSuccess())
        {
            Trace.WriteInfo(TraceComponent, "PagingDescription::FromPublicApi error: {0}", error);
            return error;
        }

        queryPagingDescription_ = make_unique<QueryPagingDescription>(move(pagingDescription));
    }

    return ErrorCode::Success();
}

void DeployedNetworkQueryDescription::GetQueryArgumentMap(__out QueryArgumentMap & argMap) const
{
    // NodeName is a mandatory query parameter.
    argMap.Insert(
        QueryResourceProperties::Node::Name,
        nodeName_);

    if (queryPagingDescription_ != nullptr)
    {
        queryPagingDescription_->SetQueryArguments(argMap);
    }
}

File: src/prod/src/managed/Api/src/System/Fabric/Description/DeployedNetworkQueryDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{    
    using System;
    using System.Fabric.Common;
    using System.Fabric.Interop;

    /// <summary>
    /// <para>Represents the query input used by <see cref="System.Fabric.FabricClient.NetworkManagementClient.GetDeployedNetworkListAsync(DeployedNetworkQueryDescription)" />.</para>
    /// </summary> 
    public sealed class DeployedNetworkQueryDescription : PagedQueryDescriptionBase
    {
        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Description.DeployedNetworkQueryDescription" /> class.</para>
        /// </summary>
        public DeployedNetworkQueryDescription()
        {            
            this.NodeName = null;
            this.ContinuationToken = null;            
        }

        /// <summary>
        /// Gets or sets the node name to query for.        
        /// </summary>
        public string NodeName { get; set; }

        internal static void Validate(DeployedNetworkQueryDescription queryDescription)
        {
            Requires.Argument<string>("NodeName", queryDescription.NodeName).NotNullOrWhiteSpace();
            ValidatePaging(queryDescription);
        }

        internal IntPtr ToNative(PinCollection pinCollection)
        {
            var nativeDescription = new NativeTypes.FABRIC_DEPLOYED_NETWORK_QUERY_DESCRIPTION();
            nativeDescription.NodeName = pinCollection.AddObject(this.NodeName);            
            nativeDescription.PagingDescription = this.ToNativePagingDescription(pinCollection);
            nativeDescription.Reserved = IntPtr.Zero;

            return pinCollection.AddBlittable(nativeDescription);
        }
    }
}
